File: ts-src/transcriber.ts
import pLimit from "p-limit";
import { getAudioInfo, splitAudio, type AudioChunk } from "./audio.js";
import { getConfig, type ASRConfig } from "./config.js";
import { APIError } from "./errors.js";

export type ContextMode = "none" | "sliding" | "full_serial";

export interface TranscriptionSegment {
  start: number;
  end: number;
  text: string;
}

export interface TranscriptionResult {
  text: string;
  segments: TranscriptionSegment[];
  duration: number;
  stats: {
    chunks: number;
    chunksTranscribed: number;
    chunksSkippedSilent: number;
    totalTime: number;
    mode: string;
    concurrency: number;
  };
}

interface APIResponseSegment {
  id: number;
  start: number;
  end: number;
  text: string;
}

interface APIResponse {
  text?: string;
  segments?: APIResponseSegment[];
  // Fallback: chat completions format
  choices?: Array<{ message?: { content?: string } }>;
  content?: string;
}

export class ASRTranscriber {
  private config: ASRConfig;

  constructor(config?: ASRConfig) {
    this.config = config ?? getConfig();
  }

  /**
   * Transcribe a single audio chunk via the /audio/transcriptions endpoint.
   * Sends multipart/form-data with the audio file buffer.
   */
  private async transcribeChunk(
    chunk: AudioChunk,
  ): Promise<{ text: string; segments: TranscriptionSegment[] }> {
    const formData = new FormData();
    const blob = new Blob([chunk.data], { type: "audio/wav" });
    formData.append("file", blob, "audio.wav");
    formData.append("model", this.config.model);

    for (let attempt = 0; attempt <= this.config.maxRetries; attempt++) {
      try {
        const controller = new AbortController();
        const timeoutId = setTimeout(
          () => controller.abort(),
          this.config.requestTimeout * 1000,
        );

        const response = await fetch(this.config.apiBase, {
          method: "POST",
          headers: {
            Authorization: `Bearer ${this.config.apiKey}`,
          },
          body: formData,
          signal: controller.signal,
        });

        clearTimeout(timeoutId);

        if (!response.ok) {
          throw APIError.fromResponse(response.status, response.statusText);
        }

        const result = (await response.json()) as APIResponse;

        // New /audio/transcriptions format: { text, segments }
        if (result.text !== undefined) {
          const segments: TranscriptionSegment[] = (result.segments ?? []).map(
            (seg) => ({
              start: seg.start,
              end: seg.end,
              text: seg.text,
            }),
          );
          return { text: result.text, segments };
        }

        // Fallback: chat completions format
        if (result.choices?.[0]?.message?.content) {
          return { text: result.choices[0].message.content, segments: [] };
        }
        if (result.content) {
          return { text: result.content, segments: [] };
        }

        return { text: "", segments: [] };
      } catch (error) {
        const normalizedError =
          error instanceof APIError
            ? error
            : error instanceof Error && error.name === "AbortError"
              ? APIError.timeout(this.config.requestTimeout)
              : new APIError(String(error), undefined, true, error);

        const isLastAttempt = attempt === this.config.maxRetries;
        const retryable = normalizedError.retryable;

        if (!retryable) {
          throw normalizedError;
        }

        if (isLastAttempt) {
          throw APIError.maxRetriesExceeded(
            this.config.maxRetries + 1,
            normalizedError,
          );
        }

        await new Promise((resolve) =>
          setTimeout(resolve, 1000 * (attempt + 1)),
        );
      }
    }

    return { text: "", segments: [] };
  }

  async transcribe(
    audioPath: string,
    contextMode: ContextMode = "sliding",
    maxConcurrency?: number,
  ): Promise<TranscriptionResult> {
    const startTime = Date.now();
    const info = await getAudioInfo(audioPath);
    const totalDuration = info.duration;

    const chunks = await splitAudio(audioPath, this.config.maxChunkDuration);

    if (chunks.length === 0) {
      return {
        text: "",
        segments: [],
        duration: totalDuration,
        stats: {
          chunks: 0,
          chunksTranscribed: 0,
          chunksSkippedSilent: 0,
          totalTime: 0,
          mode: contextMode,
          concurrency: 0,
        },
      };
    }

    const concurrency = maxConcurrency ?? this.config.maxConcurrency;
    const limit = pLimit(concurrency);
    const textResults = new Map<number, string>();
    const segmentResults = new Map<number, TranscriptionSegment[]>();

    const nonSilentChunks = chunks.filter((c) => !c.isSilent);
    const skippedSilent = chunks.length - nonSilentChunks.length;

    for (const chunk of chunks) {
      if (chunk.isSilent) {
        textResults.set(chunk.index, "");
        segmentResults.set(chunk.index, []);
      }
    }

    // New API doesn't use context passing, so all modes effectively run in parallel
    await Promise.all(
      nonSilentChunks.map((chunk) =>
        limit(async () => {
          const result = await this.transcribeChunk(chunk);
          textResults.set(chunk.index, result.text);
          // Offset API segments by the chunk's start time
          const offsetSegments = result.segments.map((seg) => ({
            start: seg.start + chunk.startMs / 1000,
            end: seg.end + chunk.startMs / 1000,
            text: seg.text,
          }));
          segmentResults.set(chunk.index, offsetSegments);
        }),
      ),
    );

    // Merge segments in chunk order
    const allSegments: TranscriptionSegment[] = [];
    for (const chunk of chunks) {
      const segs = segmentResults.get(chunk.index) ?? [];
      if (segs.length > 0) {
        allSegments.push(...segs);
      } else {
        // Fallback: create a segment from the text result
        const text = textResults.get(chunk.index) ?? "";
        if (text) {
          allSegments.push({
            start: chunk.startMs / 1000,
            end: chunk.endMs / 1000,
            text,
          });
        }
      }
    }

    const fullText = chunks
      .map((chunk) => textResults.get(chunk.index) ?? "")
      .join("");
    const elapsed = (Date.now() - startTime) / 1000;

    return {
      text: fullText,
      segments: allSegments,
      duration: totalDuration,
      stats: {
        chunks: chunks.length,
        chunksTranscribed: chunks.length - skippedSilent,
        chunksSkippedSilent: skippedSilent,
        totalTime: Math.round(elapsed * 100) / 100,
        mode: contextMode,
        concurrency,
      },
    };
  }
}

File: src/autoglm_asr_mcp/transcriber.py
"""Core ASR transcription logic using Zhipu AI audio transcription API."""

import asyncio
import time
from dataclasses import dataclass, field
from enum import Enum
from pathlib import Path
from typing import Any

import httpx

from .audio_utils import AudioChunk, get_audio_format, load_audio, split_audio_on_silence
from .config import ASRConfig, get_config


class ContextMode(str, Enum):
    NONE = "none"
    SLIDING = "sliding"
    FULL_SERIAL = "full_serial"


@dataclass
class TranscriptionSegment:
    start: float
    end: float
    text: str


@dataclass
class TranscriptionResult:
    text: str
    segments: list[TranscriptionSegment]
    duration: float
    stats: dict = field(default_factory=dict)


class ASRTranscriber:
    def __init__(self, config: ASRConfig | None = None):
        self.config = config or get_config()
        self.config.validate()
        self._client: httpx.AsyncClient | None = None
    
    async def _get_client(self) -> httpx.AsyncClient:
        if self._client is None:
            self._client = httpx.AsyncClient(timeout=self.config.request_timeout)
        return self._client
    
    async def close(self) -> None:
        if self._client:
            await self._client.aclose()
            self._client = None
    
    async def _transcribe_chunk(
        self,
        chunk: AudioChunk,
        audio_format: str = "wav",
    ) -> dict[str, Any]:
        """Transcribe a single audio chunk via /audio/transcriptions endpoint.
        
        Sends multipart/form-data with the audio file.
        Returns dict with 'text' and 'segments' keys.
        """
        mime_type = f"audio/{audio_format}"
        filename = f"audio.{audio_format}"
        
        headers = {
            "Authorization": f"Bearer {self.config.api_key}",
        }
        
        client = await self._get_client()
        
        for attempt in range(self.config.max_retries + 1):
            try:
                files = {"file": (filename, chunk.data, mime_type)}
                data = {"model": self.config.model}
                
                response = await client.post(
                    self.config.api_base,
                    headers=headers,
                    files=files,
                    data=data,
                )
                response.raise_for_status()
                result = response.json()
                
                # New /audio/transcriptions format: { text, segments }
                if isinstance(result, dict) and "text" in result:
                    segments = []
                    for seg in result.get("segments", []):
                        segments.append(TranscriptionSegment(
                            start=seg.get("start", 0),
                            end=seg.get("end", 0),
                            text=seg.get("text", ""),
                        ))
                    return {"text": result["text"], "segments": segments}
                
                # Fallback: chat completions format
                if isinstance(result, dict) and result.get("choices"):
                    text = result["choices"][0]["message"].get("content", "") or ""
                    return {"text": text, "segments": []}
                if isinstance(result, dict) and result.get("content"):
                    return {"text": result.get("content", ""), "segments": []}
                
                return {"text": "", "segments": []}
                
            except httpx.HTTPStatusError as e:
                if attempt == self.config.max_retries:
                    raise RuntimeError(f"ASR API error after {self.config.max_retries + 1} attempts: {e}")
                await asyncio.sleep(1 * (attempt + 1))
            except httpx.TimeoutException:
                if attempt == self.config.max_retries:
                    raise RuntimeError(f"ASR API timeout after {self.config.max_retries + 1} attempts")
                await asyncio.sleep(1 * (attempt + 1))
        
        return {"text": "", "segments": []}
    
    async def transcribe(
        self,
        audio_path: str | Path,
        context_mode: ContextMode = ContextMode.SLIDING,
        max_concurrency: int | None = None,
    ) -> TranscriptionResult:
        """Transcribe audio file with automatic chunking and parallel requests.
        
        The new /audio/transcriptions API doesn't require context passing,
        so all chunks are transcribed in parallel regardless of context_mode.
        The context_mode parameter is kept for backward compatibility.
        """
        start_time = time.time()
        audio_path = Path(audio_path)
        audio_format = get_audio_format(audio_path)
        
        audio = load_audio(audio_path)
        total_duration = len(audio) / 1000.0
        
        chunks = split_audio_on_silence(
            audio,
            max_chunk_duration_ms=self.config.max_chunk_duration * 1000,
        )
        
        if not chunks:
            return TranscriptionResult(
                text="",
                segments=[],
                duration=total_duration,
                stats={"chunks": 0, "total_time": 0, "mode": context_mode.value},
            )
        
        concurrency = max_concurrency or self.config.max_concurrency
        semaphore = asyncio.Semaphore(concurrency)
        text_results: dict[int, str] = {}
        segment_results: dict[int, list[TranscriptionSegment]] = {}
        skipped_silent = 0
        
        non_silent_chunks = [c for c in chunks if not c.is_silent]
        skipped_silent = len(chunks) - len(non_silent_chunks)
        
        for chunk in chunks:
            if chunk.is_silent:
                text_results[chunk.index] = ""
                segment_results[chunk.index] = []
        
        async def transcribe_with_semaphore(chunk: AudioChunk) -> None:
            async with semaphore:
                result = await self._transcribe_chunk(chunk, audio_format=audio_format)
                text_results[chunk.index] = result["text"]
                # Offset API segments by the chunk's start time
                offset_segments = []
                for seg in result["segments"]:
                    offset_segments.append(TranscriptionSegment(
                        start=seg.start + chunk.start_ms / 1000.0,
                        end=seg.end + chunk.start_ms / 1000.0,
                        text=seg.text,
                    ))
                segment_results[chunk.index] = offset_segments
        
        # All chunks run in parallel (new API doesn't need context passing)
        tasks = [transcribe_with_semaphore(chunk) for chunk in non_silent_chunks]
        await asyncio.gather(*tasks)
        
        # Merge segments in chunk order
        all_segments: list[TranscriptionSegment] = []
        for chunk in chunks:
            segs = segment_results.get(chunk.index, [])
            if segs:
                all_segments.extend(segs)
            else:
                # Fallback: create a segment from the text result
                text = text_results.get(chunk.index, "")
                if text:
                    all_segments.append(TranscriptionSegment(
                        start=chunk.start_ms / 1000.0,
                        end=chunk.end_ms / 1000.0,
                        text=text,
                    ))
        
        full_text = "".join(text_results.get(chunk.index, "") for chunk in chunks)
        elapsed = time.time() - start_time
        
        return TranscriptionResult(
            text=full_text,
            segments=all_segments,
            duration=total_duration,
            stats={
                "chunks": len(chunks),
                "chunks_transcribed": len(chunks) - skipped_silent,
                "chunks_skipped_silent": skipped_silent,
                "total_time": round(elapsed, 2),
                "mode": context_mode.value,
                "concurrency": concurrency,
            },
        )
